File: speed.py
"""!
@file speed.py
@brief This file contains the functions used to measure the performance of the E-DES algorithm.
@author Ana Vidal (118408)
@author Simão Andrade (118345)
@date 2023-10-20
"""

import e_des
import time
from Crypto.Cipher import DES

def generate_random_data(number_of_bytes : int) -> bytearray:
    """!
    @brief This function generates random data.

    @param number_of_bytes The number of bytes to generate.

    @return The random data.
    """

    urandom_file = open('/dev/urandom', 'rb')
    random_data = urandom_file.read(number_of_bytes)
    urandom_file.close()

    return random_data

def speed_encrypt(number_of_tests : bytearray, number_of_bytes : int) -> None:
    """!
    This function measures the encryption time of DES-ECB and E-DES.

    @param number_of_tests The number of tests to run.
    @param number_of_bytes The number of bytes to encrypt.

    @return None
    """


    random_data = generate_random_data(number_of_bytes)
    password = generate_random_data(e_des.BLOCK_SIZE)

    # DES-ECB encryption time
    time_list_ecb = []
    for _ in range(number_of_tests):
        start_time = time.time_ns()
        cipher = DES.new(password, DES.MODE_ECB)
        cipher.encrypt(random_data)
        end_time = time.time_ns()

        time_list_ecb.append(end_time - start_time)
    
    # E-DES encryption time
    sboxes = e_des.generate_sboxes(password)

    time_list_e_des = []
    for _ in range(number_of_tests):
        start_time = time.time_ns()
        for block_index in range(0, number_of_bytes, e_des.BLOCK_SIZE):
            block = random_data[block_index:block_index + e_des.BLOCK_SIZE]
            block = e_des.feistel_network(block, sboxes)
        end_time = time.time_ns()

        time_list_e_des.append(end_time - start_time)

    print("Number of tests: {}".format(number_of_tests))
    print("Number of bytes: {}".format(number_of_bytes))

    print("DES-ECB encryption time (in nanoseconds):")
    print("Minimum: {}".format(min(time_list_ecb)))
    print("Maximum: {}".format(max(time_list_ecb)))
    print("Average: {}".format(sum(time_list_ecb) / len(time_list_ecb)))

    print("E-DES encryption time (in nanoseconds):")
    print("Minimum: {}".format(min(time_list_e_des)))
    print("Maximum: {}".format(max(time_list_e_des)))
    print("Average: {}".format(sum(time_list_e_des) / len(time_list_e_des)))


def speed_decrypt(number_of_tests : bytearray, number_of_bytes : int) -> None:
    """!
    @brief This function measures the decryption time of DES-ECB and E-DES.

    @param number_of_tests The number of tests to run.
    @param number_of_bytes The number of bytes to decrypt.

    @return None
    """

    random_data = generate_random_data(number_of_bytes)
    password = generate_random_data(e_des.BLOCK_SIZE)

    # DES-ECB decryption time
    time_list_ecb = []
    for _ in range(number_of_tests):
        cipher = DES.new(password, DES.MODE_ECB)
        ciphertext = cipher.encrypt(random_data)

        start_time = time.time_ns()
        cipher.decrypt(ciphertext)
        end_time = time.time_ns()

        time_list_ecb.append(end_time - start_time)

    # E-DES decryption time
    sboxes = e_des.generate_sboxes(password)

    time_list_e_des = []
    for _ in range(number_of_tests):
        ciphertext = e_des.encrypt(random_data, password)

        start_time = time.time_ns()
        for block_index in range(0, number_of_bytes, e_des.BLOCK_SIZE):
            block = ciphertext[block_index:block_index + e_des.BLOCK_SIZE]
            block = e_des.inverse_feistel_network(block, sboxes)
        end_time = time.time_ns()

        time_list_e_des.append(end_time - start_time)

    
    # Print results
    print("Number of tests: {}".format(number_of_tests))
    print("Number of bytes: {}".format(number_of_bytes))

    print("DES-ECB decryption time (in nanoseconds):")
    print("Minimum: {}".format(min(time_list_ecb)))
    print("Maximum: {}".format(max(time_list_ecb)))
    print("Average: {}".format(sum(time_list_ecb) / len(time_list_ecb)))

    print("E-DES decryption time (in nanoseconds):")
    print("Minimum: {}".format(min(time_list_e_des)))
    print("Maximum: {}".format(max(time_list_e_des)))
    print("Average: {}".format(sum(time_list_e_des) / len(time_list_e_des)))

if __name__ == "__main__":

    speed_encrypt(e_des.NUMBER_OF_RUNS, e_des.BUFFER_SIZE)
    speed_decrypt(e_des.NUMBER_OF_RUNS, e_des.BUFFER_SIZE)

File: speed.c
#include "implementation.h"

/**
 * @file speed.c
 * @brief Tests the performance of the encryption and decryption functions (ECB and E-DES implementations)
 *
 * @author Ana Vidal (118408)
 * @author Simão Andrade (118345)
 * @date 2023-10-20
 */


/**
 * Function that generates random data using '/dev/urandom' and stores it in a buffer
 * 
 * @param buffer pointer to the buffer (uint8_t array)
 * @param buffer_size the buffer size (size_t)
 */
void generate_random_data(uint8_t *buffer, size_t buffer_size)
{
    FILE *urandom = fopen("/dev/urandom", "r");
    fread(buffer, sizeof(uint8_t), buffer_size, urandom);
    fclose(urandom);
}

/**
 * Function that tests the speed of the encryption functions (ECB and E-DES implementations)
 * 
 * @param number_of_tests number of tests to run (int)
 * @param number_of_bytes number of bytes to encrypt/decrypt (size_t)
 */
void speed_encrypt(const int number_of_tests, const size_t number_of_bytes)
{

    // Time variables
    struct timespec start, end;

    // Generate the random bytes
    uint8_t random_bytes[number_of_bytes];
    generate_random_data(random_bytes, number_of_bytes);

    // Generate the password
    uint8_t password[BLOCK_SIZE];
    generate_random_data(password, BLOCK_SIZE);

    /* DES-ECB CIPHER */
    printf("DES-ECB CIPHER\n");

    // Generate the key schedule
    DES_key_schedule key_schedule;
    DES_set_key((DES_cblock*)password, &key_schedule);


    clock_t *time_list_ecb = (clock_t*)malloc(number_of_tests * sizeof(clock_t));
    for (int test = 0; test < number_of_tests; test++) {
        clock_t start_time = clock();
        // Perform DES-ECB encryption here
        for (int block_index = 0; block_index < number_of_bytes; block_index += BLOCK_SIZE)
        {
            // Perform DES-ECB encryption here
            DES_ecb_encrypt((DES_cblock *) (random_bytes + block_index), (DES_cblock *) (random_bytes + block_index), &key_schedule, DES_ENCRYPT);
        }
        clock_t end_time = clock();

        time_list_ecb[test] = end_time - start_time;
    }

    // Print the results (Min, Max, Average) in milliseconds
    clock_t minimum_time_ecb = time_list_ecb[0];
    clock_t maximum_time_ecb = time_list_ecb[0];
    clock_t total_time_ecb = 0;
    for (int test = 0; test < number_of_tests; test++) {
        if (time_list_ecb[test] < minimum_time_ecb) {
            minimum_time_ecb = time_list_ecb[test];
        }
        if (time_list_ecb[test] > maximum_time_ecb) {
            maximum_time_ecb = time_list_ecb[test];
        }
        total_time_ecb += time_list_ecb[test];
    }

    printf("Minium: %f ms\n", (double)minimum_time_ecb / (CLOCKS_PER_SEC / 1000));
    printf("Maximum: %f ms\n", (double)maximum_time_ecb / (CLOCKS_PER_SEC / 1000));
    printf("Average: %f ms\n", (double)total_time_ecb / (CLOCKS_PER_SEC / 1000) / number_of_tests);

    free(time_list_ecb);

    /* E-DES CIPHER */
    printf("E-DES CIPHER\n");

    struct s_box * sboxes = (struct s_box *)malloc(NUMBER_OF_S_BOXES * sizeof(struct s_box));

    if (sboxes == NULL) {
        printf("Error allocating memory for sboxes\n");
        exit(1);
    }

    generate_sboxes(password, sboxes);

    clock_t *time_list_edes = (clock_t*)malloc(number_of_tests * sizeof(clock_t));
    uint8_t *cipher_block = (uint8_t *)malloc(BLOCK_SIZE * sizeof(uint8_t));

    if (time_list_edes == NULL || cipher_block == NULL) {
        printf("Memory allocation error\n");
        exit(1);
    }

    for (int test = 0; test < number_of_tests; test++) {
        clock_t start_time = clock();
        for (int block_index = 0; block_index < number_of_bytes; block_index += BLOCK_SIZE)
        {
            // TODO: Add feistel_network(const uint8_t *block, const struct s_box *sboxes, uint8_t **cipher_block)
        }
        clock_t end_time = clock();

        time_list_edes[test] = end_time - start_time;
    }

    free(cipher_block);
    free(sboxes);

    // Print the results (Min, Max, Average) in milliseconds
    clock_t minimum_time_edes = time_list_edes[0];
    clock_t maximum_time_edes = time_list_edes[0];
    clock_t total_time_edes = 0;
    for (int test = 0; test < number_of_tests; test++) {
        if (time_list_edes[test] < minimum_time_edes) {
            minimum_time_edes = time_list_edes[test];
        }
        if (time_list_edes[test] > maximum_time_edes) {
            maximum_time_edes = time_list_edes[test];
        }
        total_time_edes += time_list_edes[test];
    }

    printf("Minium: %f ms\n", (double)minimum_time_edes / (CLOCKS_PER_SEC / 1000));
    printf("Maximum: %f ms\n", (double)maximum_time_edes / (CLOCKS_PER_SEC / 1000));
    printf("Average: %f ms\n", (double)total_time_edes / (CLOCKS_PER_SEC / 1000) / number_of_tests);
}

/**
 * Function that tests the speed of the decryption functions (ECB and E-DES implementations)
 * 
 * @param number_of_tests number of tests to run (int)
 * @param number_of_bytes number of bytes to encrypt/decrypt (size_t)
 */
void speed_decrypt(const int number_of_tests, const size_t number_of_bytes)
{
    // Time variables
    struct timespec start, end;

    // Generate the random bytes
    uint8_t random_bytes[number_of_bytes];
    generate_random_data(random_bytes, number_of_bytes);

    // Generate the password
    uint8_t password[BLOCK_SIZE];
    generate_random_data(password, BLOCK_SIZE);

    /* DES-ECB DECIPHER */
    printf("DES-ECB DECIPHER\n");

    // Generate the key schedule
    DES_key_schedule key_schedule;
    DES_set_key((DES_cblock*)password, &key_schedule);


    clock_t *time_list_ecb = (clock_t*)malloc(number_of_tests * sizeof(clock_t));
    for (int test = 0; test < number_of_tests; test++) {
        clock_t start_time = clock();
        // Perform DES-ECB encryption here
        for (int block_index = 0; block_index < number_of_bytes; block_index += BLOCK_SIZE)
        {
            // Perform DES-ECB encryption here
            DES_ecb_encrypt((DES_cblock *) (random_bytes + block_index), (DES_cblock *) (random_bytes + block_index), &key_schedule, DES_ENCRYPT);
        }
        clock_t end_time = clock();

        time_list_ecb[test] = end_time - start_time;
    }

    // Print the results (Min, Max, Average) in milliseconds
    clock_t minimum_time_ecb = time_list_ecb[0];
    clock_t maximum_time_ecb = time_list_ecb[0];
    clock_t total_time_ecb = 0;
    for (int test = 0; test < number_of_tests; test++) {
        if (time_list_ecb[test] < minimum_time_ecb) {
            minimum_time_ecb = time_list_ecb[test];
        }
        if (time_list_ecb[test] > maximum_time_ecb) {
            maximum_time_ecb = time_list_ecb[test];
        }
        total_time_ecb += time_list_ecb[test];
    }

    printf("Minium: %f ms\n", (double)minimum_time_ecb / (CLOCKS_PER_SEC / 1000));
    printf("Maximum: %f ms\n", (double)maximum_time_ecb / (CLOCKS_PER_SEC / 1000));
    printf("Average: %f ms\n", (double)total_time_ecb / (CLOCKS_PER_SEC / 1000) / number_of_tests);

    free(time_list_ecb);

    /* E-DES DECIPHER */
    printf("E-DES DECIPHER\n");

    struct s_box * sboxes = (struct s_box *)malloc(NUMBER_OF_S_BOXES * sizeof(struct s_box));

    if (sboxes == NULL) {
        printf("Error allocating memory for sboxes\n");
        exit(1);
    }

    generate_sboxes(password, sboxes);

    clock_t *time_list_edes = (clock_t*)malloc(number_of_tests * sizeof(clock_t));
    uint8_t *cipher_block = (uint8_t *)malloc(BLOCK_SIZE * sizeof(uint8_t));

    if (time_list_edes == NULL || cipher_block == NULL) {
        printf("Memory allocation error\n");
        exit(1);
    }

    for (int test = 0; test < number_of_tests; test++) {
        clock_t start_time = clock();
        for (int block_index = 0; block_index < number_of_bytes; block_index += BLOCK_SIZE)
        {
            // TODO: Add inverse_feistel_network(const uint8_t *block, const struct s_box *sboxes, uint8_t **cipher_block)
        }
        clock_t end_time = clock();

        time_list_edes[test] = end_time - start_time;
    }

    free(cipher_block);
    free(sboxes);

    // Print the results (Min, Max, Average) in milliseconds
    clock_t minimum_time_edes = time_list_edes[0];
    clock_t maximum_time_edes = time_list_edes[0];
    clock_t total_time_edes = 0;
    for (int test = 0; test < number_of_tests; test++) {
        if (time_list_edes[test] < minimum_time_edes) {
            minimum_time_edes = time_list_edes[test];
        }
        if (time_list_edes[test] > maximum_time_edes) {
            maximum_time_edes = time_list_edes[test];
        }
        total_time_edes += time_list_edes[test];
    }

    printf("Minium: %f ms\n", (double)minimum_time_edes / (CLOCKS_PER_SEC / 1000));
    printf("Maximum: %f ms\n", (double)maximum_time_edes / (CLOCKS_PER_SEC / 1000));
    printf("Average: %f ms\n", (double)total_time_edes / (CLOCKS_PER_SEC / 1000) / number_of_tests);
}



int main(void){
    speed_encrypt(NUMBER_OF_TESTS, BUFFER_SIZE); // TODO : FIX THIS FUNCTION

    // speed_decrypt(NUMBER_OF_TESTS, BUFFER_SIZE);

    return 0;
}